infrastructure/src/test/java/org/corfudb/infrastructure/log/statetransfer/batchprocessor/StateTransferBatchProcessorTest.java

```diff
@@ -0,0 +1,141 @@
+        List<Long> addresses = LongStream.range(0L, 10L).boxed().collect(Collectors.toList());
+        List<LogData> stubList = createStubList(addresses);
+        LogUnitClient logUnitClient = mock(LogUnitClient.class);
+        AddressSpaceView addressSpaceView = mock(AddressSpaceView.class);
+        doThrow(new OverwriteException(SAME_DATA)).when(logUnitClient).writeRange(stubList);
+        ProtocolBatchProcessor batchProcessor = ProtocolBatchProcessor
+                .builder()
+                .logUnitClient(logUnitClient)
+                .addressSpaceView(addressSpaceView)
+                .build();
+        assertThatThrownBy(() -> batchProcessor.writeRecords(ReadBatch.builder().data(stubList).build(),
+                logUnitClient, new AtomicInteger(1), Duration.ofMillis(100)))
+                .isInstanceOf(OverwriteException.class);
+
+    }
+
+    @Test
+    void writeRecordsRetry() {
+        // Try writing 10 records, write only 5 of them, catch the exception.
+        // Then retry writing the rest and succeed.
+        // At the end return a transfer batch response with status SUCCEEDED and the correct
+        // initial addresses.
+        List<Long> addresses = LongStream.range(0L, 10L).boxed().collect(Collectors.toList());
+        List<LogData> stubList = createStubList(addresses);
+        List<LogData> knownStubList = stubList.stream()
+                .filter(d -> d.getGlobalAddress() % 2 != 0).collect(Collectors.toList());
+        List<Long> known = addresses.stream().filter(a -> a % 2 == 0).collect(Collectors.toList());
+        LogUnitClient logUnitClient = mock(LogUnitClient.class);
+        AddressSpaceView addressSpaceView = mock(AddressSpaceView.class);
+        CompletableFuture<Boolean> failedFuture = new CompletableFuture<>();
+        failedFuture.completeExceptionally(new OverwriteException(SAME_DATA));
+
+        doReturn(failedFuture).when(logUnitClient).writeRange(stubList);
+
+        doReturn(CompletableFuture.completedFuture(true)).when(logUnitClient)
+                .writeRange(knownStubList);
+        // Even addresses are written only
+        doReturn(CompletableFuture.completedFuture(new KnownAddressResponse(new HashSet<>(known))))
+                .when(logUnitClient)
+                .requestKnownAddresses(0L, 9L);
+
+        ProtocolBatchProcessor batchProcessor = ProtocolBatchProcessor
+                .builder()
+                .logUnitClient(logUnitClient)
+                .addressSpaceView(addressSpaceView)
+                .build();
+
+        TransferBatchResponse resp =
+                batchProcessor.writeRecords(ReadBatch.builder().data(stubList).build(),
+                        logUnitClient, new AtomicInteger(2), Duration.ofMillis(100));
+        // Should succeed and carry the initial addresses after the retry
+        assertThat(resp.getStatus()).isEqualTo(SUCCEEDED);
+        assertThat(resp.getTransferBatchRequest().getAddresses()).isEqualTo(addresses);
+    }
+
+    @Test
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://app.codacy.com/app/annymartinez.m/CorfuDB/pullRequest?prid=4651453)